### PR TITLE
Allows JSON to be passed if JSON Content-Type

### PR DESCRIPTION
**Pull request by LDonoughe**
Doesn't support hashes in general

Re: https://github.com/bblimke/webmock/issues/449

### CODE REVIEW

**Comment by bblimke** [comment]
@LDonoughe thank you for the pull request!

This would indeed be a very useful addition to Webmock.

In order to support hashes as stubbed response body, Webmock would either need
to convert them to json or make sure all http client libs support hash as a response body.
I believe this has been raised in https://github.com/bblimke/webmock/issues/449

To make the pull request is complete, an appropriate acceptance spec would be useful,
to make sure that this functionality actually works across all http client libs,
and updated README.

If we allow WebMock to accept hashes just for JSON, then perhaps
Webmock should provide a more specific error message in case Hash is provided,
but json content type is not?

**Comment by LDonoughe** [comment]
You're Welcome @bblimke !

I've tried to address your PR comments in my latest commit. The reason so many files were touched is because a number of the adaptors were relying on implicit string conversion where Hash does not have one, so I'm explicitly `to_s` on the body where I got errors. Let me know if these changes do not completely address your concerns and what I can do to change that.

File: lib/webmock/http_lib_adapters/typhoeus_hydra_adapter.rb [hunk 1]
```diff
@@ -82,8 +82,8 @@ def self.build_request_signature(req)
         def self.build_webmock_response(typhoeus_response)
           webmock_response = WebMock::Response.new
           webmock_response.status = [typhoeus_response.code, typhoeus_response.status_message]
-          webmock_response.body = typhoeus_response.body
           webmock_response.headers = typhoeus_response.headers
+          webmock_response.body = typhoeus_response.body
```
Comment (LDonoughe) [inline lib/webmock/http_lib_adapters/typhoeus_hydra_adapter.rb]: This was necessary due to the body validation relying on headers for Hash now

**Comment by LDonoughe** [comment]
@bblimke is there anything else I can do to speed up the acceptance/merge of this PR? My team could really use this functionality to properly test 3rd-party endpoints

**Comment by bblimke** [comment]
@LDonoughe sorry. I missed your latest changes.

Putting duplication aside, I'm not sure I understand the need for `to_s` change in all adapters.

Are you trying to convert the Hash object to String in order for http clients to accept it?
Hash `to_s` is not JSON.

Please see my above comment:

>In order to support hashes as stubbed response body, Webmock would either need
to convert them to json or make sure all http client libs support hash as a response body.
I believe this has been raised in #449 

The current specs also don't actually check the body returned from request.

**Comment by LDonoughe** [comment]
@bblimke I know it's been a couple years but I wanted to share my findings and close this PR.

The short answer is that I was wrong about needing to return a hash rather than a string. So, my advice to anyone who finds this PR is to check what data is actually being returned by your endpoint via your choice of HTTP adapter. If it's actually a hash and not a string, feel free to resurrect my "fix." In most cases, you're going to have to JSON.parse the string to get your JSON.

As far as the integration tests go: when I attempted to add integration tests for the JSON work I noticed that it was broken in 4 or so HTTP adapters. When I went to add a similar shared test for Array to see why that was the case I noticed that that too was broken. I lost the original code/tests I had but I've created a new Array integration test if you'd like to see for yourself by checking out my branch (https://github.com/LDonoughe/webmock/tree/data_type_support_tests) and running `bundle exec rake spec`

I sincerely hope this saves at least one person some time and effort testing their integrations.

**Comment by bblimke** [comment]
@LDonoughe thank you for these insights and the for closing the issue. Hopefully that will help others.

Yes, there are many trade-offs in WebMock in order to support various http clients,
but that's the price to have a unified API that works everywhere (in theory ;)

People can always mock the specific http clients in case WebMock is too generic.

**Pull request closed without merging**